src/Language/Mimsa/Backend/NormaliseConstructors.hs

```diff
@@ -41,6 +41,7 @@ normaliseConstructors dt expr' =
 getNestedTyCons :: Expr Name ann -> BackendM ann TyCon
 getNestedTyCons (MyConsApp _ a _) = getNestedTyCons a
 getNestedTyCons (MyConstructor _ tyCon) = pure tyCon
+getNestedTyCons (MyLambda _ _ a) = getNestedTyCons a
```

Comment:
This stops complicated types exploding - I suspect there are way more cases here that will need adding.